### PR TITLE
Fix abs paths with pkg_config generator

### LINKED ISSUE #2756: Invalid formatting of libdir/includedir in pkg_config generator [bug]
Version: conan-1.2.3
OS: Linux

https://github.com/conan-io/conan/blob/ac25c8994878706bac23efd009ae74bc294c1add/conans/client/generators/pkg_config.py#L28-L36

`pkg_config` generator doesn't check if libdir or includedir are absolute paths which leads to the following (invalid) `.pc` being generated for `boost_math/1.66.0@bincrafters/stable`:

```
prefix=/home/user/.conan/data/boost_math/1.66.0/bincrafters/stable/package/937af278b1c9d86290dddaf1322dc99fb7d75862
libdir=${prefix}//home/user/.conan/data/boost_level8group/1.66.0/bincrafters/stable/package/afc149253b16bde7d64e527e0dbfa27b203330c8/math/lib
includedir=${prefix}//home/user/.conan/data/boost_level8group/1.66.0/bincrafters/stable/package/afc149253b16bde7d64e527e0dbfa27b203330c8/math/include
...
```

### PR DESCRIPTION
**Pull request by danimtb**
- [x] Refer to the issue that supports this Pull Request: closes #2756
- [x] If the issue has missing info, explain the purpose/use case/pain/need that covers this Pull Request.
- [x] I've read the [Contributing guide](https://raw.githubusercontent.com/conan-io/conan/develop/.github/CONTRIBUTING.md).
- [x] I've followed the PEP8 style guides for Python code.
- [x] I've opened another PR in the Conan docs repo to the ``develop`` branch, documenting this one. Also adding a description of the changes in the ``changelog.rst`` file. https://github.com/conan-io/docs

### CODE REVIEW

File: conans/test/generators/pkg_config_test.py [hunk 1]
```diff
@@ -55,3 +58,94 @@ def variables_setup_test(self):
...
+[env]
+"""
+        conanfile = """
+import os
+from conans import ConanFile
+
+class PkgConfigConan(ConanFile):
+    name = "MyLib"
+    version = "0.1"
+    settings = "os", "compiler", "arch", "build_type"
```
Comment (lasote) [inline conans/test/generators/pkg_config_test.py]: Why don't you remove the settings and the profile? the generator seems to don't care about any setting. This way the package id is the same too.

File: conans/client/generators/pkg_config.py [hunk 2]
```diff
@@ -18,54 +20,69 @@
...
+            ret["%s.pc" % depname] = self.single_pc_file_contents(depname, cpp_info)
         return ret
+
+    def single_pc_file_contents(self, name, cpp_info):
+        lines = ['prefix=%s' % cpp_info.rootpath.replace("\\", "/")]
+        libdir_vars = []
+        for i, libdir in enumerate(cpp_info.libdirs):
+            libdir = libdir.replace("\\", "/")
+            varname = "libdir" if i == 0 else "libdir%d" % (i + 2)
+            prefix = "${prefix}/" if self._with_prefix(libdir) else ""
```
Comment (lasote) [inline conans/client/generators/pkg_config.py]: Having abs paths non-prefixes makes no sense here. The prefix is to be able to replace the base path and reuse the package on another computer. So, is "libdirs" finally giving abs paths? how could we get relative ones? If the user is appending an absolute one in the recipe, we need to do:

`libdir = os.path.relpath(cpp_info.rootpath, libdir)`

File: conans/client/generators/pkg_config.py [hunk 3]
```diff
@@ -18,54 +20,69 @@
...
+
+        if cpp_info.public_deps:
+            public_deps = " ".join(cpp_info.public_deps)
+            lines.append("Requires: %s" % public_deps)
+        return "\n".join(lines) + "\n"
+
+    def _concat_if_not_empty(self, groups):
+        return " ".join([param for group in groups for param in group if param and param.strip()])
+
+    def _with_prefix(self, path):
```
Comment (lasote) [inline conans/client/generators/pkg_config.py]: I think you could get rid of this function, only one line, and the name is confusing. And by the commented above, it makes no sense to check the abs path.

File: conans/test/generators/pkg_config_test.py [hunk 4]
```diff
@@ -55,3 +58,94 @@ def variables_setup_test(self):
...
+    name = "MyLib"
+    version = "0.1"
+
+    def package_info(self):
+        self.cpp_info.includedirs = []
+        self.cpp_info.libdirs = []
+        self.cpp_info.bindirs = []
+        self.cpp_info.libs = []
+        libname = "mylib"
+        fake_dir = os.path.abspath(os.path.join("my_absoulte_path", "fake"))
```
Comment (lasote) [inline conans/test/generators/pkg_config_test.py]: The test is ok, but this package will be un-reusable. 
The thing you have to test is: If you get an absolute path but built from the `self.package_folder` is prefixed ok in the pkg file, never containing an abs path there.

File: conans/test/generators/pkg_config_test.py (hunk 1 above)
Comment (danimtb) [reply conans/test/generators/pkg_config_test.py]: rpaths are only used with gcc and clang:
https://github.com/danimtb/conan/blob/37144b9a4718ce6e8a22aa7bbf35df82cbe19052/conans/client/build/compiler_flags.py#L16#23

File: conans/client/generators/pkg_config.py (hunk 2 above)
Comment (danimtb) [reply conans/client/generators/pkg_config.py]: Yes, you are right. However if you check the issue you will see that `boost_math` package has lib dirs pointing to a different path (actually pointing to the lib path of another package) and `cpp_info.rootpath` is not a prefix of libdir path.

And I think this also happens in Windows when `short_paths=True`.

**Pull request merged**